### ISSUE TITLE
Allow objects to have custom labels

### ISSUE THREAD

**Issue by Symbianx** [tigera/operator#1385]
Looks like it's impossible right now to create the `calico-system` namespace with a custom label.

## Expected Behavior

Tigera operator should merge the labels of an existing namespace (or any other resource) with the wanted labels.

## Current Behavior

Tigera operator ignores labels in the existing namespace, overriding them with the wanted ones.

## Possible Solution

Annotations are already merged in the `mergeState` util function, we could easily do the same for labels: https://github.com/tigera/operator/blob/756210bcace3e961f3ba67f8e795fd78617f8167/pkg/controller/utils/component.go#L214-L217

## Steps to Reproduce (for bugs)

1. Create the `calico-system` namespace
2. Label the namespace with `openpolicyagent.org/webhook: ignore`
3. Install tigera-operator and wait for it to start creating resources
4. The label `openpolicyagent.org/webhook` is gone.

## Context

We're trying to install the calico using the tigera operator. However, tigera is having issues creating deployments in the `calico-system` namespace because we have OPA running and configured to block deployment creations without the `app.kubernetes.io/name` label.
We were already expecting this to happen and created the `calico-system` namespace beforehand with the label `openpolicyagent.org/webhook: ignore` (namespaces labeled bypass the OPA admission webhook).

However, that label is removedby the tigera-operator.

## Your Environment

* AWS Configuration (policy only mode)
* OPA running

**Comment by Symbianx** [comment]
If the possible solution is accepted by the maintainers, I can provide a PR for it.

**Comment by tmjd** [comment]
I think I'd like to see the operator updated to add the `app.kubernetes.io/name` since that is the root of why you are looking to merge labels on the namespace.
I think that label is the direction that k8s is moving so it would be good to make that change.
@caseydavenport WDYT? Is there any reason we wouldn't want to start including that new label? 
I assume we'd keep the `k8s-app` we already have and add `app.kubernetes.io/name`.

**Comment by Symbianx** [comment]
I think the operator having the standard labels definetely makes sense and should be there.
I also think we should still give the ability for people to label the `calico-system` namespace. 

In my case, I want to disable OPA for the calico namespace, it's good practice to disable OPA for core system namespace like `kube-system`. The calico namespace manages internal cluster networking which makes it very much a core namespace and could cause whole cluster failures if OPA starts denieing API requests for it. For instance, a misconfigured OPA policy could deny new `calico-node` pods to start up which would not be good.

**Comment by caseydavenport** [comment]
@tmjd I don't see any reason not to copy the labels in our merge logic the same way we do with annotations - am I missing something there?

Definitely agree about adding some conventional app labels as well.

**Comment by kanor1306** [comment]
Is there any update about this issue? I am running into it in a similar use case. We label namespaces to enable and disable network policies. As we cannot label the namespace the calico-kube-controller to kube-api communication, ironically, is blocked by the network policies. 

We have tried to pre-create the namespace, as well as adding the label on the fly, but it keeps disappearing  (I assume that removed by the operator reconciliation cycle)

**Comment by caseydavenport** [comment]
I'm not aware of anyone currently working on this, although the change should be rather straightforward.

**Comment by tmjd** [comment]
Sorry I missed responding before to
>@tmjd I don't see any reason not to copy the labels in our merge logic the same way we do with annotations - am I missing something there?

Yeah I don't know any reason to not do that. Sounds reasonable.

I also am not aware of anyone working on it.

**Comment by sazzle2611** [comment]
We would love this feature too please so that we can disable istio sidecar injection on the namespace

Thanks :)

**Comment by lmm** [comment]
Hey all,

We've put up an outline of proposed changes to operator component configuration. Among other fields, this will include allowing adding extra labels and annotations. Please take a look. We'd appreciate your input on the proposed changes: https://github.com/tigera/operator/issues/1990

**Comment by caseydavenport** [comment]
For posterity:

- This PR added the ability to `kubectl label` any object and have it not overwritten by the operator: https://github.com/tigera/operator/pull/1680
- This PR allows users to set a variety of custom fields at install time: https://github.com/tigera/operator/pull/2063

I'm going to close this for now - if there is a specific scenario not covered by those two, please open another issue and we'll track it there.
